blueocean-events/src/main/java/io/jenkins/blueocean/events/PipelineEventListener.java

```diff
@@ -90,45 +89,14 @@ public void onNewHead(FlowNode flowNode) {
         }
     }
 
-    private List<String> getBranch(FlowNode flowNode) {
-        List<String> branch = new ArrayList<>();
-        FlowNode parentBlock = getParentBlock(flowNode);
-
-        while (parentBlock != null) {
-            branch.add(0, parentBlock.getId());
-            parentBlock = getParentBlock(parentBlock);
-        }
-
-        return branch;
-    }
-
-    private FlowNode getParentBlock(FlowNode flowNode) {
-        List<FlowNode> parents = flowNode.getParents();
-
-        for (FlowNode parent : parents) {
-            if (parent instanceof StepStartNode) {
-                if (parent.getAction(BodyInvocationAction.class) != null) {
-                    return parent;
-                }
-            }
-        }
-
-        for (FlowNode parent : parents) {
-            if (parent instanceof StepEndNode) {
-                continue;
-            }
-            FlowNode grandparent = getParentBlock(parent);
-            if (grandparent != null) {
-                return grandparent;
-            }
-        }
-
-        return null;
+    /* package: so that we can unit test it */ List<String> getBranch(FlowNode flowNode) {
+        return Lists.reverse(flowNode.getEnclosingBlocks().stream()
+                .map(FlowNode::getId).collect(Collectors.toList()));
```

Comment:
Oh hey, if you just wanted the IDs there's `node.getAllEnclosingIds()`

Please don't kill me :)  I've got APIs to do pretty much all the things you'd want to do with FlowNodes at this point...

Reply:
Ha ha. This is great. Let me make this change.